flows/microstrategy/mstro_to_s3.py

```diff
@@ -0,0 +1,144 @@
+# To find report ID, go to the report in Microstrategy then:
+## Go to Tools > Report Details Page or Document Details Page.
+## Click Show Advanced Details button at the bottom
+# Report name should be unique as it is the file name in the S3 bucket
+REPORTS = [["2020 Bond Expenses Obligated", "85A9E0A049F06D98AF1CF3BE8CDA9394"]]
+
+## AWS Credentials
+AWS_ACCESS_ID = environment_variables["AWS_ACCESS_ID"]
+AWS_PASS = environment_variables["AWS_PASS"]
+BUCKET_NAME = environment_variables["BUCKET"]
+
+# Report names and ids separated from 2D list
+report_names = [i[0] for i in REPORTS]
+report_ids = [i[1] for i in REPORTS]
+
+# Returns a connection object for interacting with the microstrategy API
+@task(
+    max_retries=2,
+    retry_delay=datetime.timedelta(minutes=10),
+    # state_handlers=[handler],
+)
+def connect_to_mstro():
+    conn = Connection(
+        base_url=BASE_URL,
+        username=MSTRO_USERNAME,
+        password=MSTRO_PASSWORD,
+        project_id=PROJECT_ID,
+        login_mode=1,
+    )
+    return conn
+
+
+# Returns a s3 object for interacting with the boto3 AWS S3 API
+@task(
+    max_retries=2,
+    retry_delay=datetime.timedelta(minutes=10),
+    # state_handlers=[handler],
+)
+def connect_to_AWS():
+    session = boto3.Session(
+        aws_access_key_id=AWS_ACCESS_ID, aws_secret_access_key=AWS_PASS,
+    )
+    s3_res = session.resource("s3")
+
+    return s3_res
+
+
+# Downloads a report from microstrategy with a given report_id
+# returns it as a pandas dataframe
+@task(
+    max_retries=2,
+    retry_delay=datetime.timedelta(minutes=10),
+    # state_handlers=[handler],
+)
+def download_report(report_id, conn):
+    my_report = Report(connection=conn, id=report_id, parallel=False)
+    return my_report.to_dataframe()
+
+
+# Takes a pandas dataframe and formats it to be sent as a .csv in an S3 bucket
+# Uses the report_name.csv as a file name
+# report_name should be unique or it'll overwrite another report
+@task(
+    max_retries=2,
+    retry_delay=datetime.timedelta(minutes=10),
+    # state_handlers=[handler],
+)
+def report_to_s3(df, report_name, s3):
+    csv_buffer = StringIO()
+    df.to_csv(csv_buffer, index=False)
+    file_name = f"{report_name}.csv"
+    s3.Object(BUCKET_NAME, file_name).put(Body=csv_buffer.getvalue())
+
+
+# Define prefect flow
+with Flow(
+    f"microstrategy_to_s3_{ENV}",
+    schedule=None,  # initalizing with no schedule
+    # DockerRun needs a docker agent
+    run_config=DockerRun(labels=[ENV, "docker", "atd-data03"]),
+) as flow:
+    # Mapping parameters
+    ids = Parameter("report_ids", default=report_ids, required=True)
+    names = Parameter("report_names", default=report_names, required=True)
+    flow.add_task(names)
+
+    # 1. Get microstrategy connection
+    conn = connect_to_mstro()
+
+    # 2. Connect using boto3 to our S3
+    s3 = connect_to_AWS()
+
+    # 3. Download report to df (for each report_id)
+    df = download_report.map(ids, unmapped(conn))
```

Comment:
if we add more reports to this script, do we really want to process multiple reports in go? might be better to take the report name, or id or whatever as an argument so that they can be scheduled separately and not fail together?

Reply:
Prefect does this a little strange, `ids` looks like a list of `ids` but the flow is behaving more like a loop where each `id` is sent to the task individually. It's called "[mapping](https://docs-v1.prefect.io/core/concepts/mapping.html)". If a `download_report` task fails, it should continue onto to the next report and not run any subsequent tasks for that report.